### ISSUE TITLE
`jax.jit(..., static_argnums=...)` keeps strong (not weak) reference in cache, causing memory leak

### ISSUE THREAD

**Issue by patrick-kidger** [google/jax#16226]
### Description

```python
import gc
import weakref
import jax

class A:
	pass

def f(x, y):
	return x

f = jax.jit(f, static_argnums=1)
f(1, A)
ref = weakref.ref(A)
del A
gc.collect()
print(ref)  # <weakref at 0x780ddf5062c0; to 'type' at 0x12ff5d0 (A)>
```
Commenting out the `f(1, A)` line instead produces `<weakref at 0x7a4f2bdce360; dead>`.

This use-case turns up in tests, when creating and destroying many JIT'd functions.

### What jax/jaxlib version are you using?

4.11

**Comment by mattjj** [comment]
It _could_ be argued that this is working as intended: caching *in general* is based on static _values_, not static object identities. That's what lets us get a cache hit on evaluating the expression `jit(f, static_argnums=0)(n())` twice where `n = lambda: 256 + (lambda: 1)()` (I had to embellish that to defeat some CPython optimizations). With caching based on values, in general whether an object with a particular value goes out of scope doesn't matter, since we can usually construct another object with the same value and thus the cache entry isn't dead. (Cache entries are dead, and hence "leaked", if they become unreachable.)

This particular example doesn't fit neatly into that mold because `A` inherits the `type` hashable definition, where equality is based on object id. That is, after deleting `A` it's _hard_ to construct another object with the same value and thus hard to get a cache hit. It's hard, but technically not impossible: Python can [reuse object ids](https://docs.python.org/3/library/functions.html#id). So, _technically_ the cache entry is not totally dead; it's just hard to reach, and perhaps impossible to reach deterministically without using a C extension. (This is probably a good reason not to put equality-by-object-id instances into cache keys if the cache can outlive the object instances, as I think [CPython really just checks pointer value](https://github.com/python/cpython/blob/ef300937c2a1b3ebe19c2835f3b46585825c1e1f/Objects/object.c#L795) and not anything else.)

Anyway, pedantry aside, I agree this is probably an undesirable behavior for objects where equality is based on object id, and we may want to revise it. But (presumably) we don't want to break the behavior for other objects, like (non-interned) ints or strings or tuples or frozen dataclasses, where equal values can be reconstructed (and hence cache entries hit) even when the original object goes out of scope. How do we tell the difference?

I'm not sure how to do it in a complete and sound way. One heuristic might be to check for each static argument `x` whether `x.__eq__ is object.__eq__`, and if so only hold a weak reference. 

WDYT?

**Comment by patrick-kidger** [comment]
> One heuristic might be to check for each static argument x whether `x.__eq__ is object.__eq__`, and if so only hold a weak reference.

Agreed -- I think this something like this is probably a reasonable heuristic, and that a heuristic is probably as good as it gets.

Two other important heuristic cases that come to mind:

- If `type(x) is types.FunctionType`. In practice I *think* it's possible to create equal nonidentical functions (equal `__code__` etc.) but in practice that's weird and functions are my most common example of a static argument with this problem.
- Nested examples of the above, e.g. `(<something with object.__eq__>,)`. (Or I suppose any pytree, with at least one heuristic-matching leaf.)
